background tasks in Python using task queues

divyanshutomar opened the pull request:
Article on running background tasks in Python using task queues

divyanshutomar committed:
added post content and images

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+---
+# Running background tasks in Python using task queues
+
+We often come across problems in our applications where a compute-intensive time-taking task needs to be performed on the server in response to some user activity or request. For a server-side application exposing a REST API, handling this problem is different from common CRUD endpoints where request-response lifecycle is usually short. In this case, the response for such a request may not be ava ...
```
Rather than starting with 'We often come across problems...", it would be more powerful to give an example where this occurs.

sidhanthp replied:
Additionally, I think this example is a little convoluted for the 1st paragraph. 

For a server-side application exposing a REST API, handling this problem is different from common CRUD endpoints where request-response lifecycle is usually short. In this case, the response for such a request may not be available, as it may be not viable to proceed with its execution immediately on the same process.
-->
When dealing with server-side applications, the response from a REST API may not be available immediately. It would be more efficient to run background tasks than waste the idle CPU cycles.

divyanshutomar replied:
It makes sense, I will rework on the introduction

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+
+We often come across problems in our applications where a compute-intensive time-taking task needs to be performed on the server in response to some user activity or request. For a server-side application exposing a REST API, handling this problem is different from common CRUD endpoints where request-response lifecycle is usually short. In this case, the response for such a request may not be ava ...
+
+The execution of such tasks or jobs can be performed in the background by some another process spawned for the sole purpose. These processes are usually called _workers_. They run concurrently with the main process (web server in our case) handling client requests. To list of all the tasks which need to be executed, a job/task queue is maintained to store tasks along with its metadata created by  ...
```
I assume you meant to write "by another process".

sidhanthp replied:
"To list all the tasks"

divyanshutomar replied:
Thanks for pointing out the grammatical mistakes. 👍

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+
+![architectire of web server and queue](./images/running-background-tasks-python/small-archi.png)
+
+Task queues are quite popular among microservices architecture. They enable each microservice to perform its definite task really well and take care of the complexities of inter-microservice communication.
```
I'm not sure what you mean when you say 'definite task'. If I were to guess, you're referring to the idea that each microservice should perform a small task.

Additionally, if you make a claim that a task queue takes 'care of the complexities of inter-microservice communication', I would explain it a little further (unless I'm missing something).

divyanshutomar replied:
I think it's not about a small or big task. I feel 'dedicated task' will make more sense here?

I will add more details here explaining the role of messaging queues in microservices architecture.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+![health check](./images/running-background-tasks-python/health-check.png)
+
+
+### Getting to Know Starter Application
```
I would use `this` to make the tool names stand out, rather than *this*

divyanshutomar replied:
Noted

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+
+### Writing the parser
+
+Let's start with writing a simple parser that accepts a Goodreads book page URL. We will be using requests python library for making an HTTP request to get HTML content of the page. BeautifulSoup is a python library that lets us search, manipulate and create structured markup languages such as HTML, XML, etc. It will create a searchable tree from the fetched page's HTML. This will allow us to ret ...
```
I would leave out "It will create a searchable tree from the fetched page's HTML."

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+
+## Real World Application
+
+We will be writing a flask-based web application which uses Redis Queue (RQ) python library for executing asynchronous tasks. The web server exposes an endpoint that accepts Goodreads book URLs. A function will parse this URL for meta information of the book like title, author, rating, etc. The function will execute asynchronously by a separate worker process. It requires a Redis server as a mess ...
```
I would include a brief explanation of why this is a good case for using RQ, since the time the process would be waiting for a response from the server is significantly more than the time spent processing.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+
+generate_redis_key_for_book = lambda bookURL: 'GOODREADS_BOOKS_INFO:' + bookURL
+
+def parse_and_persist_book_info(bookUrl):
```
I wouldn't say I'm left with a strong understanding of Redis and what it can do. If this is the bulk of your code surrounding Redis, I would go more in-depth and explain what you're doing, why you're doing it, and what's happening internally. 

It feels like I'm just copying & pasting code w/o understanding it. Not a bad idea when we're working on the web scraping, but if we're trying to teach a person how to Redis Queue's, we should explain what's going on when we're writing the code.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+    bookLinksArray = [x for x in list(set(bodyJSON)) if x.startswith('https://www.goodreads.com/book/show/')]
+    if (len(bookLinksArray)):
+      for bookUrl in bookLinksArray:
+        bookInfoParserQueue.enqueue_call(func=parse_and_persist_book_info,args=(bookUrl,),job_id=bookUrl)
```
Explain this further.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+  bookURL = request.args.get('url', None)
+  if (bookURL and bookURL.startswith('https://www.goodreads.com/book/show/')):
+    redisKey = generate_redis_key_for_book(bookURL)
+    cachedValue = redisClient.get(redisKey)
```
Explain what's going on here further.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+
+### Inspecting task queue
+
+The creators of Redis queue (RQ) library have developed another library for checking the state of Redis queue. It is called *rq-dashboard* and it can be integrated with our flask web application. It exposes a flask blueprint for integrating with an existing flask project. It's a browser-based application which shows queues status,  workers listening on those queues and jobs queued along with thei ...
```
I think you can make this a little more concise and give a brief description of the benefits.

sidhanthp replied:
It looks like you've written about it a little below, I think it might be better to move that up.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+
+Now, we are all set to begin testing our application with some Goodreads URLs. Let's start by making a POST request to `/parseGoodReads` endpoint. Make sure to provide a valid list of URLs in an array as the request body.
+
+![post request](./images/running-background-tasks-python/post-req.png)
```
What are you using to make the post request?

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+## Conclusion and Takeaways
+
+The above application demonstrates how queuing frameworks like Redis Queue can be leveraged for solving problems that need more time and computing resources. In our application, if the parsing task is executed on the same process where client requests are being processed and served, it can easily become a performance bottleneck with high traffic. The above approach not only helps in avoiding that ...
+* Queueing frameworks allows more granular control over scaling of different processes. More worker processes can be spawned if there is an accumulation of a large number of tasks in the queue.
```
"frameworks allow more"

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+
+The above application demonstrates how queuing frameworks like Redis Queue can be leveraged for solving problems that need more time and computing resources. In our application, if the parsing task is executed on the same process where client requests are being processed and served, it can easily become a performance bottleneck with high traffic. The above approach not only helps in avoiding that ...
+* Queueing frameworks allows more granular control over scaling of different processes. More worker processes can be spawned if there is an accumulation of a large number of tasks in the queue.
+* Multiple queues can be used for handling different type of tasks.
```
Did you talk about this earlier in the post?

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,161 @@
+* Multiple queues can be used for handling different type of tasks.
+* Every task can send some meta information about its status or progress so far to Redis. This information can be useful for getting an insight into a task that runs for a long duration.
+
+Thanks for following along and I hope this post would have been useful for you.
```
I don't think this line adds anything.

sidhanthp reviewed (changes requested):
I really enjoyed the style of the post and think you did a great job with a starter application, but I don't think the reader is left with a strong understanding of Redis Queue's, how to use them, and what's happening internally. 

I left you a couple comments, take a look at them and let me know what you think.

divyanshutomar committed:
reworked the article and fixed typos

divyanshutomar commented:
@sidhanthp I have made the required changes. Kindly review and let me know if anything else needs to be worked upon.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,185 @@
+---
+# Running background tasks in Python using task queues
+
+Let's say we have an e-commerce website where users can place orders for various products. Now there's a business requirement of finding out the kind of orders being placed and the most in-demand product in real-time. Also, for every order placed, the buyer should be conveyed of the confirmation of the order via an email or notification through a messaging service. In this case,if we process the  ...
```
I think this is too wordy. 

Before writing, I would think through what you want the reader to take away from the paragraph. As the first paragraph, you introduce the reader to the e-commerce website, but they've clicked on the article to learn about 'background tasks in Python' and I think it would be more effective starting with that. You only touch about the importance of background tasks at the end of the paragraph. 

Imagine I'm a reader who doesn't know what background tasks are / why I would want them, how would you introduce me with as little background?

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,185 @@
+
+Task or message queues are quite popular among microservices architecture. They enable each microservice to perform its dedicated task and work as a medium for inter-microservice communication. These queues store messages or data incoming from _producer_ microservices which can be processed or consumed by _consumer_ microservices. In the e-commerce example above, the REST API handling orders is a ...
+
+## Queueing frameworks to the rescue
```
I think this section would be more effective at the bottom. First, teach the reader why they need background tasks, about RQ, then point them to the other solutions.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,185 @@
+
+Let's say we have an e-commerce website where users can place orders for various products. Now there's a business requirement of finding out the kind of orders being placed and the most in-demand product in real-time. Also, for every order placed, the buyer should be conveyed of the confirmation of the order via an email or notification through a messaging service. In this case,if we process the  ...
+
+The execution of such tasks or jobs can be performed in the background by another process spawned for the sole purpose. These processes are usually called _workers_. They run concurrently with the main process (web server in our case) handling client requests. To list all the tasks which need to be executed, a job/task queue is maintained to store tasks along with its metadata created by incoming ...
```
Try to make this more concise. Here's an example --> 

`Workers` can be used to execute these tasks in the background. They run concurrently in the background using a queue, and the worker executes the tasks chronologically. 

This modular approach prevents the web server from being blocked from responding to incoming client requests.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,185 @@
+
+## Real World Application
+
+We will be writing a flask-based web application which retrieves _Goodreads_ book information like title, author, rating and description. The web server exposes an endpoint that accepts book URLs. A function will crawl and parse this URL  for meta information of the book. As this function will take time to execute and may lead to blocking of the main thread, we will execute it asynchronously by p ...
```
Break the paragraph here. I think it makes it easier to follow. 

"""
... pushing it to Redis queue (RQ).

RQ allows us to enqueue ... 
"""

sidhanthp commented on posts/running-background-tasks-python.md:
````diff
@@ -0,0 +1,185 @@
+  return dict(title=title.strip() if title else '',author=author.strip() if author else '',rating=float(rating.strip() if rating else 0),description=description)
+```
+
+We can now write a function called `parse_and_persist_book_info` that calls the above parsing function and persists the value to Redis so that it can be retrieved later. This function along with its arguments will be pushed to queue so that the worker process can execute it. Redis is a key-value store where the key should be unique else it may lead to overwriting of a previous value. Here `genera ...
````
Introduce a paragraph break here -->

"""
... process can execute it.

Redis is a key-value ...
"""

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,185 @@
+#........
+
+# This generates a unique Redis key against a book URL
+generate_redis_key_for_book = lambda bookURL: 'GOODREADS_BOOKS_INFO:' + bookURL
```
I would explain in a little more detail what you're doing here if this is a major piece of the post.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,185 @@
+  redisKey = generate_redis_key_for_book(bookUrl) # get Redis key for given book URL 
+  bookInfo  = parse_book_link_for_meta_data(bookUrl) # get book meta information from parsing function above
+  # Set the value to Redis. Here pickle serializes the dictionary
+  redisClient.set(redisKey,pickle.dumps(bookInfo))
```
I would explain `redisClient.set(redisKey,pickle.dumps(bookInfo))` further. I could be missing it, but I don't see where you initialized `redisClient`.

sidhanthp commented on posts/running-background-tasks-python.md:
```diff
@@ -0,0 +1,185 @@
+
+### The endpoint for accepting URLs
+
+Let's set up an endpoint that will accept a list of valid Goodreads book URLs. This is going to support POST method with URLs accepted as an array in _application/json_ body format. For validating the Goodreads book URLs we check for unique occurrences of URLs which starts with the string `https://www.goodreads.com/book/show/`. After the validation check, all valid URLs are pushed to Redis queue  ...
```
I would take out the first couple sentences, and make the paragraph this -->
After the validation check, all valid URLs are pushed to Redis queue for parsing information. Here the method `enqueue_call` of Redis queue instance takes in a function that will be executed by worker process along with required arguments of the function.

I don't think the first couple sentences add to the explanation of Redis Queues.

sidhanthp reviewed (changes requested):
Great job with the post, I've left you with a few comments that I think make the article a little easier to read and digest. 

I found it difficult to follow the longer paragraphs. Some of the paragraphs are big blocks of texts, which I think causes your eyes to glaze over the text. If you break up the paragraphs, or make the takeaways obvious using _italics_, I think it would make the article a little easier to follow.

divyanshutomar committed:
updated the article text

divyanshutomar commented:
@sidhanthp Updated

sidhanthp committed:
Updated

You moved the wrong paragraph to the bottom, I fixed it.

sidhanthp committed:
updated structure and made concise

Updated structure to 'move up' information dedicated to Redis Queues and removed information that was specific to the starter application.

divyanshutomar commented:
@sidhanthp Great job with the final touches. I guess we decided to have an example of a background job in the introductory para, that's why I kept the e-commerce example. Nevertheless, if you want to keep it concise the current state looks good to me.

sidhanthp committed:
updated grammar
